tests/testapps/AddressBookSample/AddressBook/SingleContactViewController.m

```diff
@@ -0,0 +1,417 @@
+        NSData* contactPhoto =
+            (__bridge_transfer NSData*)ABPersonCopyImageDataWithFormat(originalContact, kABPersonImageFormatOriginalSize);
+        UIImage* image = [UIImage imageWithData:contactPhoto];
+        [self.contactPhoto setImage:image];
+    }
+}
+
+- (CFTypeRef)_copyContactInfoFrom:(UITextField*)textField {
+    if ([textField.text length] > 0) {
+        return (__bridge_retained CFTypeRef)textField.text;
+    } else {
+        return NULL;
+    }
+}
+
+- (IBAction)buttonPressed:(id)sender {
+    CFErrorRef error = NULL;
+    bool succeeded = true;
+    switch (self.contactOperation) {
+        case kAddressBookViewContact:
+            // Viewing a contact doesn't need to make any changes.
+            break;
+        case kAddressBookRemoveContact:
+            // Remove the contact from the AddressBook.
+            succeeded = succeeded && ABAddressBookRemoveRecord(_addressBook, self.contact, &error);
+            succeeded = succeeded && ABAddressBookSave(_addressBook, &error);
+            if (!succeeded) {
+                NSLog(@"There was an error removing the contact: %@", (__bridge_transfer NSString*)CFErrorCopyDescription(error));
+            }
+            break;
+        case kAddressBookModifyContact:
+            // Modify the AddressBook record for the current contact to reflect
+            // any changes made by the user.
+            [self _modifyContact];
+            break;
+        case kAddressBookCreateContact:
+            // Create a new record in the AddressBook for the information
+            // entered by the user.
+            [self _createContact];
+            break;
+    }
+
+    // Dismiss this view after the button is pressed.
+    [self.navigationController popViewControllerAnimated:YES];
+}
+
+- (void)_modifyContact {
+    CFErrorRef error = NULL;
+    CFTypeRef firstName = [self _copyContactInfoFrom:self.firstName];
+    CFTypeRef lastName = [self _copyContactInfoFrom:self.lastName];
+    CFTypeRef phoneNumber = [self _copyContactInfoFrom:self.phoneNumber];
+    CFTypeRef email = [self _copyContactInfoFrom:self.email];
+    bool succeeded = true;
+
+    succeeded = succeeded && ABRecordSetValue(self.contact, kABPersonFirstNameProperty, firstName, &error);
+    if (firstName) {
+        CFRelease(firstName);
+    }
+
+    succeeded = succeeded && ABRecordSetValue(self.contact, kABPersonLastNameProperty, lastName, &error);
+    if (lastName) {
+        CFRelease(lastName);
+    }
+
+    if (self.phoneNumber.isEnabled) {
+        ABMultiValueRef existingPhone = ABRecordCopyValue(self.contact, kABPersonPhoneProperty);
+        ABMutableMultiValueRef phoneNumbers = ABMultiValueCreateMutableCopy(existingPhone);
+        CFRelease(existingPhone);
+
+        // Either update the value of the first phone number (since that's the one we displayed), or
+        // remove it entirely if the user cleared the text field.
+        if (phoneNumber) {
```

Comment:
nit: I would prefer validity checks like this when the objects are assigned/created, that way you can easily exit on failure cases and don't have to have multiple checks for validity of memory throughout the code. Note: you can wrap them in CFAutoRelease, so when you exit on failure cases you don't have to have explicit releases for the ones above. If not you can write a macro to release multiple objects. Although autorelease seems more appropriate.